File: autonomy/camera-radar-fusion/src/shared/fusion/types/fused_track.py
from shared.camera.types.camera_detection import CameraDetection
from shared.radar.types.radar_detection import RadarDetection


class FusedTrack:
    """Represents a fused track combining camera and radar information."""

    def __init__(
        self,
        camera_detection: CameraDetection,
        radar_detection: RadarDetection,
        fusion_confidence: float,
    ) -> None:
        """
        Initialize fused track.

        Args:
            camera_detection: Associated camera detection
            radar_detection: Associated radar detection
            fusion_confidence: Combined confidence score (0-1)

        """
        self.camera_detection = camera_detection
        self.radar_detection = radar_detection
        self.fusion_confidence = fusion_confidence

        # Fused properties combining both sensors
        self.bounding_box = camera_detection.bounding_box
        self.class_name = camera_detection.class_name
        self.class_id = camera_detection.class_id

        # 3D position and velocity from radar
        self.position_3d = radar_detection.position_3d
        self.velocity = radar_detection.velocity
        self.distance = radar_detection.distance

        # Image coordinates (use center of bounding box)
        self.image_coordinate_x = camera_detection.center_x
        self.image_coordinate_y = camera_detection.center_y

        # Movement classification
        self.is_moving = self.velocity > 0.5  # threshold in m/s

    def __repr__(self) -> str:
        return (
            f"FusedTrack(class={self.class_name}, "
            f"distance={self.distance:.1f}m, "
            f"velocity={self.velocity:.1f}m/s, "
            f"confidence={self.fusion_confidence:.2f})"
        )

File: autonomy/camera-radar-fusion-rust/src/shared/fusion/types/fused_track.rs
use crate::config::AppConfig;
use crate::shared::camera::types::camera_detection::CameraDetection;
use crate::shared::radar::types::radar_detection::RadarDetection;
use nalgebra::Vector4;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FusedTrack {
    pub camera_detection: CameraDetection,
    pub radar_detection: RadarDetection,
    pub fusion_confidence: f64,
}

impl FusedTrack {
    pub fn new(
        camera_detection: CameraDetection,
        radar_detection: RadarDetection,
        fusion_confidence: f64,
    ) -> Self {
        Self {
            camera_detection,
            radar_detection,
            fusion_confidence,
        }
    }

    pub fn bounding_box(&self) -> Vector4<f64> {
        self.camera_detection.bounding_box
    }

    pub fn class_name(&self) -> &str {
        &self.camera_detection.class_name
    }

    pub fn velocity(&self) -> f64 {
        self.radar_detection.velocity
    }

    pub fn distance(&self) -> f64 {
        self.radar_detection.distance
    }

    pub fn image_coordinate_x(&self) -> f64 {
        self.camera_detection.center_x()
    }

    pub fn image_coordinate_y(&self) -> f64 {
        self.camera_detection.center_y()
    }

    pub fn is_moving(&self, config: &AppConfig) -> bool {
        self.velocity() > config.movement_velocity_threshold_mps
    }
}

impl std::fmt::Display for FusedTrack {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "FusedTrack(class={}, distance={:.1}m, velocity={:.1}m/s, confidence={:.2})",
            self.class_name(),
            self.distance(),
            self.velocity(),
            self.fusion_confidence
        )
    }
}
